### PR TITLE
feat: implement gRPC API and migrate influxdb_iox_client to use it

### PR DESCRIPTION
**Pull request by tustvold**
Builds on #852 and migrates the influxdb_iox_client to use this new API

### CODE REVIEW

File: data_types/src/database_rules.rs [hunk 1]
```diff
@@ -388,7 +390,7 @@ pub enum PartitionSort {
 
 impl Default for PartitionSort {
     fn default() -> Self {
-        Self::LastWriteTime
+        Self::CreatedAtTime
```
Comment (tustvold) [inline data_types/src/database_rules.rs:393]: This and the change below make the defaulting consistent with the defaulting for MutableBufferConfig. Unfortunately the gRPC conversions cannot use the MutableBufferConfig defaulting logic because there is no way to specify an explicit NULL in gRPC

File: influxdb_iox_client/src/client/flight.rs [hunk 2]
```diff
@@ -13,23 +13,56 @@ use arrow_deps::{
 use futures_util::stream::StreamExt;
 use serde::Serialize;
 use std::{convert::TryFrom, sync::Arc};
+use thiserror::Error;
 use tonic::Streaming;
 
-use crate::errors::{GrpcError, GrpcQueryError};
+/// Error responses when querying an IOx database using the Arrow Flight gRPC
```
Comment (tustvold) [inline influxdb_iox_client/src/client/flight.rs:19]: Move this error from `errors::GrpcQueryError` into the `flight` module

File: src/influxdb_ioxd/http.rs [hunk 3]
```diff
@@ -39,6 +39,32 @@ use std::{fmt::Debug, str, sync::Arc};
 mod format;
 use format::QueryOutputFormat;
 
+/// Constants used in API error codes.
```
Comment (tustvold) [inline src/influxdb_ioxd/http.rs:42]: This can be removed once the HTTP Management API is removed

File: influxdb_iox_client/src/client/management.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
+use std::num::NonZeroU32;
+
+use thiserror::Error;
+
+use self::generated_types::{management_service_client::ManagementServiceClient, *};
+
+use crate::connection::Connection;
+
+/// Re-export generated_types
+pub mod generated_types {
```
Comment (tustvold) [inline influxdb_iox_client/src/client/management.rs:10]: This re-export means that downstream crates don't need to also depend on generated_types, they can just depend on the influxdb_iox_client.

File: tests/end-to-end.rs [hunk 5]
```diff
@@ -66,27 +71,38 @@ async fn read_and_write_data() {
...
-    grpc_api::read_group_test(&http_client, &influxdb2, &mut storage_client).await;
-    grpc_api::read_window_aggregate_test(&http_client, &influxdb2, &mut storage_client).await;
+    storage_api::read_group_test(&mut management_client, &influxdb2, &mut storage_client).await;
+    storage_api::read_window_aggregate_test(
+        &mut management_client,
+        &influxdb2,
+        &mut storage_client,
+    )
+    .await;
+    management_api::test(&mut management_client).await;
```
Comment (tustvold) [inline tests/end-to-end.rs:105]: Hijacking this test is not ideal, but avoids problems due to the TestServer's being created in parallel and clashing on ports. I believe @alamb is working on a better way to do this, or if I'm mistaken on that, I will spend some time cleaning this up in a future PR.

**Comment by tustvold** [comment]
I'm marking this as Ready for review but I imagine there will be some discussion to be had on some of the implementation decisions. In particular:

* This implements the client in terms of types within the client crate and generated_types, it does not depend on data_types - this was discussed previously on #848
* This splits up the influxdb_iox_client crate with separate clients, errors, etc... per gRPC service, and a shared connection builder to avoid duplication - this was previously discussed [here](https://github.com/influxdata/influxdb_iox/pull/862#discussion_r582142763)

File: tests/end-to-end.rs (hunk 5 above)
Comment (alamb) [reply tests/end-to-end.rs:105]: I agree that Hijacking the test is not ideal, but we have been doing it for about a year so I see no reason to fix the pattern in this PR.

For completeness, what I plan is described in https://github.com/influxdata/influxdb_iox/issues/843 -- namely to upgrade the query tests to run against an actual IOx server using data files rather than hard coded tests.

I *do not* currently plan to change the `end-to-end.rs` test in this effort -- I planned to make something in parallel to end-to-end.rs that uses the management API introduced in this PR to setup and run query engine tests. 

So in other words, if you were suggesting (and had the ambition to) split up end-to-end.rs so that it could run multiple individual `#[tokio::test]` that would probably be valuable.

File: tests/end-to-end.rs [hunk 6]
```diff
@@ -379,16 +394,23 @@ impl TestServer {
...
+
+                        match health.check_storage().await {
                             Ok(_) => {
                                 println!("Storage service is running");
-                                break;
                             }
-                            Err(e) => println!("Error checking storage service status: {}", e),
+                            Err(e) => {
+                                println!("Error checking storage service status: {}", e);
+                                continue;
```
Comment (alamb) [inline tests/end-to-end.rs:412]: I don't understand the need for this `continue` -- it seems like it will skip the `interval.tick().await` all below -- aka skip the sleep?

File: tests/end-to-end.rs [hunk 7]
```diff
@@ -379,16 +394,23 @@ impl TestServer {
...
+                    Ok(connection) => {
                         println!("Successfully connected to server");
 
-                        match client.check_storage().await {
+                        let mut health = influxdb_iox_client::health::Client::new(connection);
+
+                        match health.check_storage().await {
                             Ok(_) => {
                                 println!("Storage service is running");
-                                break;
```
Comment (alamb) [inline tests/end-to-end.rs:389]: I might be missing something, but I think this should be `break` to get out of the loop when the server is ready otherwise it will just loop until the overall timeout occurs. 

when I ran `cargo test --tests -- read_and_write --nocapture` it seems like the test is waiting the full 3 seconds for the server to be up rather than starting when the server is ready (note the `WARNING: server was not ready: deadline has elapsed`): 


```
Waiting for HTTP server to be up: error sending request for url (http://127.0.0.1:8090/health): error trying to connect: tcp connect error: Connection refused (os error 61)
Waiting for gRPC API to be up: Connection error: transport error: error trying to connect: tcp connect error: Connection refused (os error 61)
Waiting for HTTP server to be up: error sending request for url (http://127.0.0.1:8090/health): error trying to connect: tcp connect error: Connection refused (os error 61)
Waiting for gRPC API to be up: Connection error: transport error: error trying to connect: tcp connect error: Connection refused (os error 61)
Feb 28 06:13:19.826  WARN influxdb_iox::influxdb_ioxd: NO PERSISTENCE: using memory for object storage
Feb 28 06:13:19.828  INFO influxdb_iox::influxdb_ioxd: gRPC server listening bind_address=127.0.0.1:8092
Feb 28 06:13:19.839  INFO influxdb_iox::influxdb_ioxd: HTTP server listening bind_address=127.0.0.1:8090
Feb 28 06:13:19.839  INFO influxdb_iox::influxdb_ioxd: InfluxDB IOx server ready git_hash="b1939270e1e4cbeb43818dd36c5a1ab6fda07b54"
Successfully connected to server
Feb 28 06:13:20.312  INFO influxdb_iox::influxdb_ioxd::http: Processing request request=Request { method: GET, uri: /health, version: HTTP/1.1, headers: {"accept": "*/*", "host": "127.0.0.1:8090"}, body: Body(Empty) }
Feb 28 06:13:20.312  INFO influxdb_iox::influxdb_ioxd::http: Successfully processed request response=Response { status: 200, version: HTTP/1.1, headers: {"x-powered-by": "Routerify v2.0.0-beta-2"}, body: Body(Full(b"OK")) }
Successfully got a response from HTTP: Response { url: Url { scheme: "http", username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(8090), path: "/health", query: None, fragment: None }, status: 200, headers: {"x-powered-by": "Routerify v2.0.0-beta-2", "content-length": "2", "date": "Sun, 28 Feb 2021 11:13:20 GMT"} }
Storage service is running
Successfully connected to server
Storage service is running
Successfully connected to server
Storage service is running
Successfully connected to server
Storage service is running
Successfully connected to server
Storage service is running
WARNING: server was not ready: deadline has elapsed
```

File: tests/end_to_end_cases/management_api.rs [hunk 8]
```diff
@@ -0,0 +1,156 @@
+use std::num::NonZeroU32;
```
Comment (alamb) [inline tests/end_to_end_cases/management_api.rs:1]: Very thorough tests.

**Review by alamb — APPROVED** [review]
This is really nice work @tustvold  -- well done.

It is really cool to the see the management API seemingly just "slide in place easily" -- even though I realize you put in a bunch of effort to lay the groundwork to make this fit in. 

I think the server ready check in end-to-end.rs probably should be fixed, but it could be done as a follow on PR too in my opinion

File: influxdb_iox_client/Cargo.toml [hunk 9]
```diff
@@ -10,12 +10,12 @@ flight = ["arrow_deps", "serde/derive", "serde_json", "futures-util"]
...
 # Workspace dependencies, in alphabetical order
 arrow_deps = { path = "../arrow_deps", optional = true }
-data_types = { path = "../data_types" }
 generated_types = { path = "../generated_types" }
 
 # Crates.io dependencies, in alphabetical order
 futures-util = { version = "0.3.1", optional = true }
-reqwest = { version = "0.11.0", features = ["gzip", "json"] }
+http = "0.2.3"
+hyper = "0.14"
```
[the remark this answers is not in the document]
Comment (tustvold) [reply influxdb_iox_client/Cargo.toml:18]: hyper is both a client and a server. The client is the basis for both reqwest and tonic. Normally you wouldn't need to use it directly, but the tonic client builder currently doesn't expose the connection timeout - https://github.com/hyperium/tonic/issues/498

File: influxdb_iox_client/src/client/management.rs [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    UnexpectedError(#[from] tonic::Status),
+}
+
+/// Result type for the management client
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// An IOx Management API client.
+///
+/// ```no_run
+/// #[tokio::main]
```
[the remark this answers is not in the document]
Comment (tustvold) [reply influxdb_iox_client/src/client/management.rs:48]: I have to come clean - I copied them from the old implementation so you probably have @domodwyer to thank for these :smile:

File: tests/end-to-end.rs (hunk 6 above)
Comment (tustvold) [reply tests/end-to-end.rs:412]: Nice catch, yeah that was a holdover from a previous version of this method that also constructed the storage client - before I realised this was redundant and not actually testing anything additional :grin:

File: tests/end-to-end.rs (hunk 7 above)
Comment (tustvold) [reply tests/end-to-end.rs:389]: Yeah - this was a refactoring fail - nice catch

File: influxdb_iox_client/src/connection.rs [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+        Self {
+            connect_timeout: timeout,
+            ..self
+        }
+    }
+
+    /// Bounds the total amount of time a single client HTTP request take before
+    /// being aborted.
+    ///
+    /// This timeout includes:
```
[the remark this answers is not in the document]
Comment (tustvold) [reply influxdb_iox_client/src/connection.rs:109]: Again - not written by myself :grinning:

**Comment by tustvold** [comment]
> Is there some test that ensures that the defaults are consistent? That is probably a good one to add so that we don't leave a latent lurking timebomb / hard to debug issue for anyone who may want to change the defaults in the future

They currently aren't because they can't be without breaking the HTTP API, I will fix this as part of removing it.

> Do you plan a subsequent PR to remove the HTTP management API? It seems like we have at least:

I've created #889 as a barebones tracking issue for the completed and remaining work

> So in other words, if you were suggesting (and had the ambition to) split up end-to-end.rs so that it could run multiple individual #[tokio::test] that would probably be valuable.

Created #890 to track. The main thing I can see being a bit of a nuisance is the log output, but we shall see

File: influxdb_iox_client/src/client/management.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (alamb) [reply influxdb_iox_client/src/client/management.rs:49]: I should have known *you* hadn't gone 🍌 's yet -- lol :)

File: influxdb_iox_client/src/client/management.rs [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+impl Client {
+    /// Creates a new client with the provided connection
+    pub fn new(channel: tonic::transport::Channel) -> Self {
+        Self {
+            inner: ManagementServiceClient::new(channel),
+        }
+    }
+
+    /// Set the server's writer ID.
+    pub async fn update_writer_id(&mut self, id: NonZeroU32) -> Result<(), Error> {
```
Comment (domodwyer) [inline influxdb_iox_client/src/client/management.rs:87]: It seems wrong that `update_writer_id()` can return "Database already exists" - it forces the client to match against a load of impossible errors (which I assume will be growing to a fairly large list).

Should we not return an enum that includes only the errors that can actually happen? It also serves as documentation - "this call can fail in these ways" which might not be apparent otherwise.

File: influxdb_iox_client/src/client/management.rs [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    /// Creates a new IOx database.
+    pub async fn create_database(&mut self, rules: DatabaseRules) -> Result<(), Error> {
+        self.inner
+            .create_database(CreateDatabaseRequest { rules: Some(rules) })
+            .await
+            .map_err(|status| match status.code() {
+                tonic::Code::AlreadyExists => Error::DatabaseAlreadyExists,
+                tonic::Code::FailedPrecondition => Error::NoWriterId,
```
Comment (domodwyer) [inline influxdb_iox_client/src/client/management.rs:121]: Do we have any tests that assert these mappings match the server side? Falling out of sync would be pretty terrible!

File: src/influxdb_ioxd/http.rs [hunk 14]
```diff
@@ -39,6 +39,32 @@ use std::{fmt::Debug, str, sync::Arc};
...
+///
+/// Expressing this as a enum prevents reuse of discriminants, and as they're
+/// effectively consts this uses UPPER_SNAKE_CASE.
+#[allow(non_camel_case_types)]
+#[derive(Debug, PartialEq)]
+pub enum ApiErrorCode {
```
Comment (domodwyer) [inline src/influxdb_ioxd/http.rs:48]: I'll be glad to see these gone!

**Review by domodwyer — CHANGES_REQUESTED** [review]
Looks good! I think it'll be much nicer to use and develop with - thank you for the hard work.

Slightly concerned we're going to wind up with one giant "every possible error" enum but I'm sure we can sort something sensible 👍

File: influxdb_iox_client/src/client/management.rs (hunk 13 above)
Comment (tustvold) [reply influxdb_iox_client/src/client/management.rs:121]: There's a test for `DatabaseAlreadyExists`, but I'll need to do something different for `NoWriterId` - I'll see what I can come up with

File: influxdb_iox_client/src/client/management.rs (hunk 12 above)
Comment (tustvold) [reply influxdb_iox_client/src/client/management.rs:87]: Whilst I agree in principle, maintaining per-method enumerations is quite an overhead... When you call a function you're only really interested in the errors that you can do something about, if `update_writer_id` returns some nonsense DatabaseAlreadyExists error you can just propagate it up along with all the other errors you don't have custom handling logic for? Yes per-method enumerations might give a better indication as to some of the failures you might encounter, but, at least in my experience, custom error recovery is generally the special case that is added when the failure scenario in question is already well understood by the author - i.e. you have a match expression that special cases the one or two variants you specifically want to handle, and if you get something else, you just bubble it up.

@shepmaster might have some thoughts here? It certainly seems to be fairly common to reuse error enumerations across modules, or even entire crates but I'm not certain what the best practice here is.

Comment (domodwyer) [reply influxdb_iox_client/src/client/management.rs:87]: It sounds like you agree, but your objection is the overhead of adding the enum is too high? In which case I'm happy to do it, I don't think defining an enum is a big deal - I'd rather we expose a precise and clear API client that doesn't push the burden to the user.

I may be wrong, but I can foresee there being many error states as the code base expands - Conductor's fine-grained visibility into IOx nodes (including metrics/errors/everything else) will be what allows it to be affective at maintaining a healthy cluster.

Comment (tustvold) [reply influxdb_iox_client/src/client/management.rs:87]: I think it is more akin to the debate around checked vs unchecked exceptions, with per-method enumerations being akin to checked exceptions and less-granular enumerations being more akin to the unchecked exceptions. I personally am firmly in the latter camp, and plenty has been written on the topic, but I will add per-method enumerations in order to move forward with this PR.

Comment (domodwyer) [reply influxdb_iox_client/src/client/management.rs:87]: I'm not sure I agree with the exception comparison (but we do agree on exception handling! exceptions are exceptional! don't be checking those things) but sounds like a plan.

Per-method enums might make the status code <-> error mapping tests easier too - if we can make the error types symmetric they can be easily checked with simple property tests which are probably easier to cover than the integration tests 👍

File: influxdb_iox_client/src/client/management.rs (hunk 13 above)
Comment (tustvold) [reply influxdb_iox_client/src/client/management.rs]: I've created #900 to track this, as the API is likely to evolve and currently does not have any customers that need to distinguish error codes, I feel it would be premature to invest significant further effort into the error handling story and its associated testing.

File: influxdb_iox_client/src/client/management.rs (hunk 12 above)
Comment (alamb) [reply influxdb_iox_client/src/client/management.rs]: I think it may be a good idea to get this client wired into the conductor and see which rough edges are the most important to hone first. It is hard for me to fully understand the specific use / concern about these potential implementations

**Review by domodwyer — APPROVED** [review]
Thanks @tustvold - looking forward to getting this going in Conductor!

**Comment by alamb** [comment]
🎉

**Comment by alamb** [comment]
Part of #942
